src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    vlog(_logger.debug, "archival_metadata_stm::sync()");
+    /*
+     * Implementation note
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
+     * changes)
+     * Case 2: we have already synced (once or more) during the current
+     * term, but need to sync again (e.g. re-starting the archiver to apply
+     * topic configs)
+     *
+     * They are handled separately to the reader's delight. In theory, it is
+     * possible to fallthrough from Case 1 to Case 2. This function should be
+     * called in a loop until it succeeds.
+     */
+
+    // If there's any replications pending, wait for them to complete.
+    // Note that we throw away the result and use the Raft's committed offset
+    // below. This deals with both success and failure paths.
+    if (_last_replicate) {
+        auto fut = std::move(_last_replicate.value());
+        _last_replicate = std::nullopt;
+        if (!fut.available()) {
+            vlog(
+              _logger.debug, "Waiting for ongoing replication before syncing");
+
+            const auto before_replication = model::timeout_clock::now();
+
+            try {
+                std::ignore = co_await ss::with_timeout(
+                  before_replication + timeout, std::move(fut));
+            } catch (const ss::timed_out_error&) {
+                vlog(_logger.error, "Replication wait timed out");
+                co_return false;
+            }
+
+            const auto after_replication = model::timeout_clock::now();
+            timeout -= after_replication - before_replication;
+        } else {
+            co_await fut.discard_result();
```

Comment:
I think here you don't need to add a scheduling point. You can just call `fut.get().discard_result()`. Call to `get()` will propagate exception if the replication has failed.

Reply:
Makes sense. I realised that we probably don't want to propagate the exception here. It's fine for replication to fail. `sync` should continue in that case.

Reply:
Fixed by using `ignore_ready_future` if the future was ready. Also updated the other branch to ignore any exceptions from `consensus::replicate` and proceed.